File: CS/PersonalOrganizer/Common/ViewModel/Messages.cs
using System;
using System.Linq;
using System.ComponentModel;

namespace PersonalOrganizer.Common.ViewModel {
    /// <summary>
    /// Represents the type of entity state change notification when the IUnitOfWork.SaveChanges method has been called.
    /// </summary>
    public enum EntityMessageType {

        /// <summary>
        /// The new entity has been added to the unit of work. 
        /// </summary>
        Added,

        /// <summary>
        /// The object has been deleted from the unit of work.
        /// </summary>
        Deleted,

        /// <summary>
        /// One of the properties on the object has been modified. 
        /// </summary>
        Changed
    }

    public class EntityMessage<TEntity, TPrimaryKey> {

        public TPrimaryKey PrimaryKey { get; private set; }

        /// <summary>
        /// The entity state change notification type.
        /// </summary>
        public EntityMessageType MessageType { get; private set; }

        public EntityMessage(TPrimaryKey primaryKey, EntityMessageType messageType) {
            this.PrimaryKey = primaryKey;
            this.MessageType = messageType;
        }
    }

    public class SaveAllMessage {
    }

    public class CloseAllMessage {
        readonly CancelEventArgs cancelEventArgs;
        public CloseAllMessage(CancelEventArgs cancelEventArgs) {
            this.cancelEventArgs = cancelEventArgs;
        }
        public bool Cancel {
            get { return cancelEventArgs.Cancel; }
            set { cancelEventArgs.Cancel = value; }
        }
    }

    public class NavigateMessage<TNavigationToken> {
        public NavigateMessage(TNavigationToken token) {
            Token = token;
        }
        public TNavigationToken Token { get; private set; }
    }
}

File: VB/PersonalOrganizer/Common/ViewModel/Messages.vb
Imports System
Imports System.Linq
Imports System.ComponentModel
Namespace PersonalOrganizer.Common.ViewModel
  ''' <summary>
  ''' Represents the type of entity state change notification when the IUnitOfWork.SaveChanges method has been called.
  ''' </summary>
  Public Enum EntityMessageType  
    ''' <summary>
    ''' The new entity has been added to the unit of work. 
    ''' </summary>
    Added    
    ''' <summary>
    ''' The object has been deleted from the unit of work.
    ''' </summary>
    Deleted    
    ''' <summary>
    ''' One of the properties on the object has been modified. 
    ''' </summary>
    Changed
  End Enum
  Public Class EntityMessage(Of TEntity, TPrimaryKey)
    Private _MessageType As EntityMessageType
    Private _PrimaryKey As TPrimaryKey
    Public ReadOnly Property PrimaryKey As TPrimaryKey
      Get
        Return _PrimaryKey
      End Get
    End Property    
    ''' <summary>
    ''' The entity state change notification type.
    ''' </summary>
    Public ReadOnly Property MessageType As EntityMessageType
      Get
        Return _MessageType
      End Get
    End Property
    Public Sub New(ByVal primaryKey As TPrimaryKey, ByVal messageType As EntityMessageType)
      Me._PrimaryKey = primaryKey
      Me._MessageType = messageType
    End Sub
  End Class
  Public Class SaveAllMessage
  End Class
  Public Class CloseAllMessage
    Private ReadOnly _cancelEventArgs As CancelEventArgs
    Public Sub New(ByVal cancelEventArgs As CancelEventArgs)
      Me._cancelEventArgs = cancelEventArgs
    End Sub
    Public Property Cancel As Boolean
      Get
        Return _cancelEventArgs.Cancel
      End Get
      Set
        _cancelEventArgs.Cancel = value
      End Set
    End Property
  End Class
  Public Class NavigateMessage(Of TNavigationToken)
    Private _Token As TNavigationToken
    Public Sub New(ByVal token As TNavigationToken)
      Me._Token = token
    End Sub
    Public ReadOnly Property Token As TNavigationToken
      Get
        Return _Token
      End Get
    End Property
  End Class
End Namespace
